JavaScript/chapter01/p06_string_compression/avc278.js

```diff
@@ -0,0 +1,54 @@
+// String Compression: Implement a method to perform basic string compression using the counts of repeated characters.
+// For example, the string aabcccccaaa would become a2b1c5a3. If the "compressed" string would not become smaller than
+// the original string, your method should return the origin string. You can assume the string has only uppercase and
+// lowercase letters (a-z).
+
+const assert = require("assert");
+
+/**
+ * Returns the compressed version of the string if applicable
+ * @param  {string}  str input string to compress
+ * @return {string}      either the compressed string, or the string
+ *
+ * Storing the outputStr is at worst size 2N for an input string whose size is N, which is still O(N) additional space.
+ * Storing and updating the pointers `idx`, `next`, and `freq` take O(1) additional space and O(1) time.
+ * As we iterate through the str in the outer while loop, in one extreme, assuming no repeating letters in a sequence,
+ * idx increments by 1 each iteration, and so it's O(N) runtime. On the other extreme, assuming all repeating letters,
+ * this would be O(N) on the outer loop, and O(N) in the inner while loop.
+ * Runtime: O(N^2)
```

Comment:
IIUC the statement in L17 is true. yet this runtime analysis is not correct. how can that be? what is the correct runtime?

(although in a weird twist of things, it turns out that the implementation as written _is_ O(N²) due to the fact that `outputStr` is being concatenated character by character. JavaScript strings being immutable causes a new string of length O(N) to be allocated in each of the O(N) rounds of the external loop, which causes the whole thing to be quadratic. the solution is to have an array of strings and join them all at the end: arrays do still need to be reallocated when they are appended, _but_ they are created with some additional capacity to avoid having to reallocate every single time a new element is appended).

Reply:
Hmm, so let me walk you through my thought process for this as it was a bit confusing, and I'm still not sure if I fully understand it:

Say you have a string "aaaaa", which gets compressed to "a5".
Let's step through the algorithm
`outputStr = '', idx = 0, next = 1`
enter while loop since idx (0) is less than str.length (5)
`currLetter = 'aaaaa'[0] = 'a', freq = 1`
enter true while loop
`'a' === 'aaaaa'[1] (true)`
so we increment next and freq `next=2, freq=2`
`'a' === 'aaaaa'[2] (true)`
so we increment next and freq `next=3, freq=3`
`'a' === 'aaaaa'[3] (true)`
so we increment next and freq `next=4, freq=4`
`'a' === 'aaaaa'[4] (true)`
so we increment next and freq `next=5, freq=5`
`'a' === 'aaaaa'[5] (false, this is actually a bug that you pointed out down below which I'll fix, but something that JS is so nice to take of for us)`
so we hit the `else` and create our string `outputStr = '' + currLetter(a)freq(5)` => 'a5'
and update idx from 0 to 0+5 = 5, and next from 5 to 6.
We go back to the start of our while loop and break since 5 !< 5.

In the inner loop for this case, we iterated 5 times (the length of the string), and in the outer loop we only iterate once. So I guess for this case, the run time was actually O(N)?

=========

Perhaps for a different example "aabb"
`outputStr = '', idx = 0, next = 1`
enter outer while loop
`currLetter = 'aabb'[0] = 'a', freq = 1`
enter true while loop
`'a' === 'aabb'[1] (true)`
so we increment next and freq `next=2, freq=2`
`'a' === 'aabb'[2] (false)
so we hit the `else` and update our string `outputStr = '' + currLetter(a)freq(2)` => `a2`
and update idx from 0 to 0+2 = 2, and next from 2 to 3.
So this is O(1) on the outer loop, but O(a) where a is the length of the substring containing connected a's.

`currLetter = 'aabb'[2] = 'b', freq = 1`
enter true while loop
`'b' === 'aabb'[3] (true)`
so we increment next and freq `next=4, freq=2`
`'b' === 'aabb'[4] (false)`
so we hit the `else` and update our string `outputStr = 'a2' + currLetter(b)freq(2)` => `a2b2`
and update idx from 2 to 2+2 = 4, and next from 4 to 5.
Again this is the second iteration on the outer loop and so I think O(d) where d is the number of distinct elements side by side, and inner loop O(b) where b is the length of the substring containing connected b's.

Overall runtime for this example was O(d * (a + b))?

Reply:
the correct runtime complexity is O(N). Normally we multiply the complexities of nested loops, _but_ in cases like this we can use a different strategy: count the number of times the `str[idx]` and `str[next]` expressions are going to be evaluated. every iteration of the inner loop it will evaluate `str[next]` exactly once (with a different `next` value every time), and every iteration of the outer loop it will evaluate `str[idx]` exactly once (with a different `idx` value every time). This means that effectively both loops are really just one loop being interleaved, which yields a final answer of O(N).

Reply:
>both loops are really just one loop being interleaved

This is exactly what I wasn't getting at first, but I understand now. Thanks! 👏